### PR TITLE
Core SDK Implementation

### PR DESCRIPTION
**Pull request by dadheech115**
This PR contains work for-
- Client Factory 
- HTTPClient
- Middleware Support
- Middlewares for Authentication and Redirect
- Network call mechanism with support for multiple request going through the pipeline simultaneously.
- And Unit Tests for all the implmentation

### CODE REVIEW

**Review by darrelmiller — APPROVED** [review]

File: MSGraphCoreSDK/MSGraphCoreSDK/Common/MSConstants.m [hunk 1]
```diff
@@ -0,0 +1,16 @@
+//
+// Copyright (c) Microsoft Corporation. All Rights Reserved. Licensed under the MIT License. See License in the project root for license information.
+//
+
+#import "MSConstants.h"
+
+NSString *const MSGraphBaseURL = @"https://graph.microsoft.com/v1.0";
+NSString *const MSHeaderSdkVersion = @"SdkVersion";
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/Common/MSConstants.m:8]: Should sdk version be "SdkVersion-0.1.0"?

File: MSGraphCoreSDK/MSGraphCoreSDK/Authentication/MSAuthenticationProvider.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+typedef void(^MSAuthenticationCompletion)(NSMutableURLRequest *request, NSError *error);
+
+/**
+ The `MSAuthenticationProvider` is a protocol that is used to inject authentication into the MSHTTPClient.
+ It should handle all initial authentication, refreshing, and returning the access token.
+ */
+@protocol MSAuthenticationProvider <NSObject>
+
+/**
+ Gets the access token. This method should be implmented by a class which should provide the capability of providing the access token.
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/Authentication/MSAuthenticationProvider.h]: minor: spelling correction in the comment  for - "implemented"

File: MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            }];
+        }
+    }
+    [task setInnerTask:sessionTask];
+    [sessionTask resume];
+}
+
+- (void)setNext:(id<MSGraphMiddleware>)nextMiddleware
+{
+    if(_nextMiddleware)
```
Comment (deepak2016) [inline MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m]: As this middleware is always going to be the last middleware, don't provide mechanism to set next middleware.
Remove this implementation?

**Review by deepak2016 — APPROVED** [review]

File: MSGraphCoreSDK/MSGraphCoreSDK/Authentication/MSAuthenticationProvider.h (hunk 2 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDK/Authentication/MSAuthenticationProvider.h]: Sure. Will correct that.

File: MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m (hunk 3 above)
Comment (dadheech115) [reply MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m]: a stub will suffice in that case I think?
Can't remove the implementation as the method is part of MSGraphMiddleware protocol and this class conforms to it.

Comment (deepak2016) [reply MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m]: Yes, stub should be fine. Maybe with some assert warning that setnext is not allowed in this middleware.

**Pull request merged**